File: src/erlang/cat.erl
-module(cat).
-import(string,[len/1]).
-export([start/0]).

read_blocks(D) ->
    case file:read(D,131072) of
	{ok, Data} ->
	    LEN = len(Data),
	    if 
		LEN > 0 ->
		    io:fwrite(Data),
		    read_blocks(D);
		LEN == 0 ->
		    file:close(D),
		    halt(0)
	    end;
	{error, Reason} ->
	    io:fwrite(Reason)
    end.

readfile(Filename) ->
    case file:open(Filename, [read]) of
	{ok,D} ->
	    read_blocks(D);
	{error,Reason} ->
	    io:fwrite(Reason)
    end.

start() ->
    readfile("../data"),
    halt(0).

File: src/rust/cat.rs
use std::fs::File;
use std::io::{Read,Write};
use std::os::unix::io::{RawFd, FromRawFd};

fn go() -> Result<(), std::io::Error> {
    let mut f = File::open("../data")?;
    let mut done = false;
    let mut buf = [0u8; 131072];
    let mut file = unsafe { File::from_raw_fd(1 as RawFd) };

    while !done {
        match f.read(&mut buf) {
            Ok(len) => {
                if len == 0 {
                    done = true;
                }
                else {
                    file.write_all(&mut buf)?;
                }
            },
            Err(_) => { done = true; }
        }
    }
    Ok(())
}

fn main() { 
    match go() {
        Ok(_) => (),
        Err(e) => {
            println!("{}",e);
        }
    }
}
